tests/test_bigquery.py

```diff
@@ -1,38 +1,74 @@
 import unittest
 import os
 import threading
+from test.support import EnvironmentVarGuard
 
 from http.server import BaseHTTPRequestHandler, HTTPServer
 
 from google.cloud import bigquery
+from google.auth.exceptions import DefaultCredentialsError
 
 HOSTNAME = "127.0.0.1"
 PORT = 8000
 URL = "http://%s:%s" % (HOSTNAME, PORT)
 
 class TestBigQuery(unittest.TestCase):
-    def test_proxy(self):
-        httpd = HTTPServer((HOSTNAME, PORT), HTTPHandler)
-        threading.Thread(target=httpd.serve_forever).start()
-        client = bigquery.Client()
-
-        try:
-            for ds in client.list_datasets(): pass
-        except:
-            pass
-
-        httpd.shutdown()
-        self.assertTrue(HTTPHandler.called, msg="Fake server did not recieve a request from the BQ client.")
-        self.assertTrue(HTTPHandler.header_found, msg="X-KAGGLE-PROXY-DATA header was missing from the BQ request.")
-
-class HTTPHandler(BaseHTTPRequestHandler):
-    called = False
-    header_found = False
-
-    def do_HEAD(s):
-        s.send_response(200)
-
-    def do_GET(s):
-        HTTPHandler.called = True
-        HTTPHandler.header_found = any(k for k in s.headers if k == "X-KAGGLE-PROXY-DATA" and s.headers[k] == "test-key")
-        s.send_response(200)
+    
+    def _test_proxy(self, client, should_use_proxy):
+        class HTTPHandler(BaseHTTPRequestHandler):
+            called = False
+            header_found = False
+
+            def do_HEAD(s):
+                s.send_response(200)
+
+            def do_GET(s):
+                HTTPHandler.called = True
+                HTTPHandler.header_found = any(k for k in s.headers if k == "X-KAGGLE-PROXY-DATA" and s.headers[k] == "test-key")
+                s.send_response(200)
+
+        with HTTPServer((HOSTNAME, PORT), HTTPHandler) as httpd:
```

Comment:
Where is this HTTP server used?  I don't see anything pointing to the address it serves.

Reply:
The server is started with the env var: KAGGLE_DATA_PROXY_URL=http://127.0.0.1:8000, so that's what the connection uses to make a call to this HTTP server.

Reply:
Mhh okay :).  I can see the server serves `127.0.0.1:8000`, but not how `KAGGLE_DATA_PROXY_URL` is set to it (so the client would actually use it), or how we test the case where it isn't used by the client if that is still relevant.  I'm probably missing where this is done -- I am still struggling with PRs...

If `KAGGLE_DATA_PROXY_URL` is indeed set to `127.0.0.1:8000`, the server address (host/port components) used in this test should be built from it.  I know this is existing, but that avoid duplication 
(reducing the chance of breakages) and make this more readable.